docs/metrics.md

```diff
@@ -0,0 +1,37 @@
+# Metrics
+
+The contents of this document may be subject to frequent change.
+It covers what each metric measures, and what information we may derive from it.
+
+## Barrier Latency
+
+Prerequisite: [Checkpoint](./checkpoint.md)
+
+This metric measures the duration from which a barrier is injected into **all** sources in the stream graph,
+to the barrier flown through all executors in the graph.
+
+### What can we understand from it?
+
+Usually when examining barrier latency, we look at **high barrier latency**.
+
+There are two contributing factors to it:
+1. Time taken to actually process the streaming messages.
+2. Buffer capacity for streaming messages.
+
+When injecting a new barrier,
+there will usually be streaming messages in the stream graph (unless it's the initial barrier).
+Since we keep total order for streaming messages,
+this means that all streaming messages currently in the stream graph have to be processed
+before the barrier can pass through.
+If barrier latency is high, it could mean a long time is taken to process these streaming messages.
+Concretely, here are some costs of processing streaming messages:
+1. CPU cost of evaluating expressions.
+2. I/O remote exchange between fragments.
+3. hash-join / hash-agg cache-miss (results in extra costs to access state on s3).
+
+Next, high barrier latency could also be caused by buffers in the graph.
```

Comment:
May give a hint that in this case we can tweak the buffer size through configuration file (developer), and a heuristic algorithm is on the way. 🤤

Reply:
> May give a hint that in this case we can tweak the buffer size through configuration file (developer),
Which options can configure this? So far I just see `stream_connector_message_buffer_size`, but that's just for connector source -> source executor.

> and a heuristic algorithm is on the way. 🤤
Any discussions / issues / RFCs I can link to?

Reply:
> Which options can configure this? So far I just see `stream_connector_message_buffer_size`, but that's just for connector source -> source executor.

https://github.com/risingwavelabs/risingwave/blob/a36e01307d60491b91870ac5a37049a378fe986f/src/config/example.toml#L49-L50

> Any discussions / issues / RFCs I can link to?

I guess we can link to #8654.

Reply:
Updated, PTAL